# Covid-19 Data Exploration

In [1]:
# pandas is a library for dataframe and data wrangling
import pandas as pd

In [2]:
# Access the dataset for the analysis later
data = pd.read_csv("C:/Users/tf199/Documents/Data Wrangling Online/Coursework 1/excess_mortality.csv")

In [3]:
# Take a look at the first 5 rows of the dataset
data.head()

location        date  p_scores_all_ages  p_scores_15_64  p_scores_65_74  \
0  Austria  2020-01-05             -10.83           -4.99          -23.21   
1  Austria  2020-01-12              -8.35           -1.45           -5.43   
2  Austria  2020-01-19               0.59           -7.32           -1.09   
3  Austria  2020-01-26               1.83           14.20           -5.11   
4  Austria  2020-02-02               7.06           -5.91           -0.22   

   p_scores_75_84  p_scores_85plus  deaths_2020_all_ages  \
0           -5.11           -11.17                1613.0   
1           -5.30           -13.40                1702.0   
2            9.77            -2.91                1797.0   
3           10.18            -5.01                1779.0   
4           19.98             5.79                1947.0   

   average_deaths_2015_2019_all_ages  deaths_2015_all_ages  ...  \
0                             1809.0                1704.0  ...   
1                             1857.0                1768.0  ...   
2                             1786.4                1837.0  ...   
3                             1747.0                1792.0  ...   
4                             1818.6                1837.0  ...   

   deaths_2017_all_ages  deaths_2018_all_ages  deaths_2019_all_ages  \
0                2293.0                1767.0                1647.0   
1                2340.0                1842.0                1709.0   
2                2060.0                1786.0                1641.0   
3                1938.0                1714.0                1713.0   
4                2045.0                1870.0                1732.0   

   deaths_2010_all_ages  deaths_2011_all_ages  deaths_2012_all_ages  \
0                1557.0                1579.0                1574.0   
1                1582.0                1671.0                1563.0   
2                1573.0                1513.0                1586.0   
3                1573.0                1560.0                1595.0   
4                1544.0                1521.0                1545.0   

   deaths_2013_all_ages  deaths_2014_all_ages  Week  deaths_2021_all_ages  
0                1637.0                1549.0     1                1944.0  
1                1659.0                1543.0     2                1941.0  
2                1612.0                1578.0     3                   NaN  
3                1584.0                1463.0     4                   NaN  
4                1752.0                1574.0     5                   NaN  

[5 rows x 21 columns]

The result from `data.head()`shows the dataset we are looking at contains mortality numbers of all age in different countries recorded in weeks as well as some statistical values (p-score). The P-score is the percentage difference between the reported number of weekly deaths (described by [GitHub:covid-19-data
](https://github.com/owid/covid-19-data/tree/master/public/data/excess_mortality))

In [4]:
# Check the basic information of the dataset
# Focus on the numbers of rows and columns, any null values in columns, and each column's data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2041 entries, 0 to 2040
Data columns (total 21 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   location                           2041 non-null   object 
 1   date                               2041 non-null   object 
 2   p_scores_all_ages                  1891 non-null   float64
 3   p_scores_15_64                     1867 non-null   float64
 4   p_scores_65_74                     1867 non-null   float64
 5   p_scores_75_84                     1867 non-null   float64
 6   p_scores_85plus                    1867 non-null   float64
 7   deaths_2020_all_ages               1837 non-null   float64
 8   average_deaths_2015_2019_all_ages  1906 non-null   float64
 9   deaths_2015_all_ages               1714 non-null   float64
 10  deaths_2016_all_ages               1872 non-null   float64
 11  deaths_2017_all_ages               1872 non-null   float

From the result of `data.info()`, we can see:
* There are 2041 rows and 21 columns
* 21 coulmns including location of where the data was collected, some statistical indicators, number of deaths in each year, Week number where the data was recorded
* Multiple columns have missing values (rows < 2024), e.g. deaths_2020_all_ages. 
* The data types of location and date are object; the datatype of week is int64; the rest of data types are float64

In [5]:
# Use data.describe() to get a statistical overview of the dataset
data.describe()

p_scores_all_ages  p_scores_15_64  p_scores_65_74  p_scores_75_84  \
count        1891.000000     1867.000000     1867.000000     1867.000000   
mean            9.190122        1.354226       11.032775        8.758945   
std            20.067535       17.171083       22.471298       23.502712   
min           -29.010000     -100.000000      -66.670000      -61.110000   
25%            -1.905000       -8.065000       -1.305000       -4.520000   
50%             4.120000       -0.890000        6.950000        3.430000   
75%            13.315000        7.755000       17.665000       15.125000   
max           156.300000      130.770000      157.560000      157.750000   

       p_scores_85plus  deaths_2020_all_ages  \
count      1867.000000           1837.000000   
mean         12.661939           5158.144801   
std          24.043579          10680.458704   
min         -48.860000             28.000000   
25%          -1.100000            835.000000   
50%           7.240000           1899.000000   
75%          19.110000           4212.000000   
max         195.050000          79012.000000   

       average_deaths_2015_2019_all_ages  deaths_2015_all_ages  \
count                        1906.000000           1714.000000   
mean                         4657.339874           4334.851225   
std                          9219.355104           9053.016442   
min                            35.800000             25.000000   
25%                           822.650000            773.000000   
50%                          1820.900000           1527.500000   
75%                          3810.700000           3339.750000   
max                         60696.400000          61731.000000   

       deaths_2016_all_ages  deaths_2017_all_ages  deaths_2018_all_ages  \
count           1872.000000           1872.000000           1872.000000   
mean            4560.031517           4688.101496           4731.537393   
std             9011.301026           9295.991808           9398.136794   
min               30.000000             31.000000             25.000000   
25%              817.000000            821.500000            821.000000   
50%             1799.000000           1794.500000           1824.500000   
75%             4373.500000           4090.000000           4315.750000   
max            57603.000000          61084.000000          67491.000000   

       deaths_2019_all_ages  deaths_2010_all_ages  deaths_2011_all_ages  \
count           1872.000000           1509.000000           1560.000000   
mean            4716.149038           2524.724321           2449.937821   
std             9396.365076           2747.142136           2707.644600   
min               30.000000             22.000000             25.000000   
25%              818.750000            838.000000            789.000000   
50%             1811.500000           1483.000000           1327.000000   
75%             4027.750000           2686.000000           2628.000000   
max            58542.000000          12967.000000          13127.000000   

       deaths_2012_all_ages  deaths_2013_all_ages  deaths_2014_all_ages  \
count           1560.000000           1612.000000           1612.000000   
mean            2511.628846           4036.009305           4033.116625   
std             2807.562006           8826.324400           8855.671307   
min               21.000000             27.000000             28.000000   
25%              777.750000            786.250000            787.000000   
50%             1359.000000           1501.500000           1489.500000   
75%             2780.250000           2955.750000           2965.250000   
max            13459.000000          60790.000000          56021.000000   

              Week  deaths_2021_all_ages  
count  2041.000000             30.000000  
mean     25.444390           3816.433333  
std      15.901442           4721.597974  
min       1.000000            369.000000  
25%      11.000000           1214.750000  
50%     

From the result above, we can see a distribution of the death occurred from different years. The min and maximum values vary from different years. The highest/max death counts was from 2020 in the middle of Covid-19 period.

## Question 1 solution
Create a DataFrame that consists of the required columns:\
location, deaths_2020_all_ages, deaths_2019_all_ages,the difference between deaths in 2020 and 2019

In [6]:
# Create a DataFrame df and extract existing columns from the original dataset ==> data
df = data[['location','deaths_2020_all_ages','deaths_2019_all_ages']]

In [7]:
# Delete rows that contain null values
df = df.dropna()

In [8]:
# Adding a column 'death_excess' to DataFrame df by deaths_2020_all_ages - deaths_2019_all_ages
df['death_excess'] = df['deaths_2020_all_ages'] - df['deaths_2019_all_ages']

In [9]:
# Take a look at DataFrame df
df.head(5)

location  deaths_2020_all_ages  deaths_2019_all_ages  death_excess
0  Austria                1613.0                1647.0         -34.0
1  Austria                1702.0                1709.0          -7.0
2  Austria                1797.0                1641.0         156.0
3  Austria                1779.0                1713.0          66.0
4  Austria                1947.0                1732.0         215.0

From above, the new dataframe results in 1817 rows and 4 fresh columns. The death_excess helps to see whether in the week there was excess death or reduced death in each location.

## Question 2 solution
Find all the countries show excess death: death_excess > 0. Excess deaths is defined as the difference between
the deaths occurred in a specific week in 2020 vs 2019. Hint: the diff between 2020 and 2019 has
to be greater than 0. 

In [10]:
dfx = data[['location','deaths_2020_all_ages','deaths_2019_all_ages','Week']]

In [11]:
# Delete rows that contain null values
dfx = dfx.dropna()

In [12]:
# Adding a column 'death_excess' to DataFrame df by deaths_2020_all_ages - deaths_2019_all_ages
dfx['death_excess'] = dfx['deaths_2020_all_ages'] - dfx['deaths_2019_all_ages']

In [13]:
# Create a subset from the DataFrame df that satiesfies where death_excess > 0
dfx = dfx.loc[dfx["death_excess"] > 0]

In [14]:
# Take a look at the subset
dfx

location  deaths_2020_all_ages  deaths_2019_all_ages  Week  \
2           Austria                1797.0                1641.0     3   
3           Austria                1779.0                1713.0     4   
4           Austria                1947.0                1732.0     5   
8           Austria                1768.0                1764.0     9   
11          Austria                1836.0                1619.0    12   
...             ...                   ...                   ...   ...   
2030  United States               65398.0               55880.0    46   
2031  United States               67517.0               56195.0    47   
2032  United States               68607.0               55458.0    48   
2033  United States               71688.0               57300.0    49   
2034  United States               73812.0               57655.0    50   

      death_excess  
2            156.0  
3             66.0  
4            215.0  
8              4.0  
11           217.0  
...            ...  
2030        9518.0  
2031       11322.0  
2032       13149.0  
2033       14388.0  
2034       16157.0  

[1176 rows x 5 columns]

In [15]:
# Find all the conuntries that have death_excess > 0
dfx['location'].unique()

array(['Austria', 'Belgium', 'Bulgaria', 'Canada', 'Chile', 'Croatia',
       'Czechia', 'Denmark', 'England & Wales', 'Estonia', 'Finland',
       'France', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Israel',
       'Italy', 'Latvia', 'Lithuania', 'Luxembourg', 'Netherlands',
       'New Zealand', 'Northern Ireland', 'Norway', 'Poland', 'Portugal',
       'Scotland', 'Slovakia', 'Slovenia', 'South Korea', 'Spain',
       'Sweden', 'Switzerland', 'Taiwan', 'United States'], dtype=object)

In [16]:
# Count the numbers of countries that have death_excess > 0
dfx['location'].nunique()

36

There are 36 countries showing excess deaths.

## Question 3 solution
For every country, find the top-5 weeks with the highest number of excess deaths.

In [17]:
# Create a new DataFrame that includes the following columns:
# location, death_excess, week
df1 = data[['location','Week','deaths_2020_all_ages','deaths_2019_all_ages']]

In [18]:
df1 = df1.dropna()

In [19]:
df1['death_excess'] = df1['deaths_2020_all_ages'] - df1['deaths_2019_all_ages']

In [20]:
# Create a subset
df_subset = df1[['location','Week','death_excess']]

In [22]:
df_subset.head(5)

location  Week  death_excess
0  Austria     1         -34.0
1  Austria     2          -7.0
2  Austria     3         156.0
3  Austria     4          66.0
4  Austria     5         215.0

In [21]:
# Group countries and sort death_excess in a descending order and extract the top 5 values for each country
# replaces the previous DataFrame index with the new integer index provided by reset_index() 
df_max = df_subset.groupby('location').apply(lambda x: x.sort_values(by='death_excess',ascending=False).head(5)).reset_index(drop=True)

In [23]:
df_max

location  Week  death_excess
0          Austria    49         948.0
1          Austria    47         859.0
2          Austria    48         855.0
3          Austria    50         807.0
4          Austria    46         727.0
..             ...   ...           ...
175  United States    15       23526.0
176  United States    16       22231.0
177  United States    17       20173.0
178  United States    50       16157.0
179  United States    14       15635.0

[180 rows x 3 columns]

In [24]:
# Take a look at the df_max
# Show all rows, the function suggested by the link
# https://dev.to/chanduthedev/how-to-display-all-rows-from-data-frame-using-pandas-dha
pd.set_option('display.max_rows', None)  
df_max

location  Week  death_excess
0             Austria    49         948.0
1             Austria    47         859.0
2             Austria    48         855.0
3             Austria    50         807.0
4             Austria    46         727.0
5             Belgium    15        2339.0
6             Belgium    14        1861.0
7             Belgium    45        1692.0
8             Belgium    16        1626.0
9             Belgium    46        1423.0
10           Bulgaria    48        2636.0
11           Bulgaria    49        2570.0
12           Bulgaria    47        2418.0
13           Bulgaria    50        2219.0
14           Bulgaria    46        1896.0
15             Canada    17        1390.0
16             Canada    18        1280.0
17             Canada    16         965.0
18             Canada    15         965.0
19             Canada    19         940.0
20              Chile    23        1575.0
21              Chile    24        1553.0
22              Chile    25        1253.0
23              Chile    22        1038.0
24              Chile    26        1013.0
25            Croatia    47         527.0
26            Croatia    48         456.0
27            Croatia    46         336.0
28            Croatia    45         314.0
29            Croatia    44         216.0
30            Czechia    44        2146.0
31            Czechia    45        2102.0
32            Czechia    46        1617.0
33            Czechia    43        1495.0
34            Czechia    47        1221.0
35            Denmark    51         231.0
36            Denmark    52         173.0
37            Denmark    15         125.0
38            Denmark    12         101.0
39            Denmark    13          90.0
40    England & Wales    16       13326.0
41    England & Wales    17       11938.0
42    England & Wales    15        8225.0
43    England & Wales    18        6746.0
44    England & Wales    14        6261.0
45            Estonia    52          56.0
46            Estonia    14          55.0
47            Estonia    31          50.0
48            Estonia    39          50.0
49            Estonia    48          47.0
50            Finland    47         143.0
51            Finland    17         141.0
52            Finland    37         133.0
53            Finland    26         126.0
54            Finland    15         122.0
55             France    14        7330.0
56             France    15        6540.0
57             France    13        4875.0
58             France    16        4108.0
59             France    45        4024.0
60            Germany    52        6266.0
61            Germany    51        4904.0
62            Germany    50        4430.0
63            Germany    33        3734.0
64            Germany    49        3186.0
65             Greece    47         933.0
66             Greece    48         900.0
67             Greece    49         819.0
68             Greece    46         568.0
69             Greece    31         328.0
70            Hungary    48        1537.0
71            Hungary    47        1502.0
72            Hungary    46        1486.0
73            Hungary    49        1341.0
74            Hungary    50        1236.0
75            Iceland    36          21.0
76            Iceland     1          18.0
77            Iceland     4          18.0
78            Iceland    21          14.0
79            Iceland     2          14.0
80             Israel    41         307.0
81             Israel    38         247.0
82             Israel    43         238.0
83             Israel    42         227.0
84             Israel    40         213.0
85              Italy    13       10908.0
86              Italy    12        9216.0
87              Italy    14        8126.0
88              Italy    15        6770.0
89              Italy    11        4844.0
90             Latvia    51         188.0
91             Latvia    27         122.0
92             Latvia    50         108.0
93             Latvia    34          96.0
94             Latvia  

From the result above, we can find the top-5 weeks with the highest number of excess deaths for each country. For example, Austria reached the top 5 highest excess deaths in Week 49, 47, 48, 50, 46 with death_excess numbers 948, 859, 855, 807 and 727 respectively. The week information is an indication of month and season in a year as well, normally Winter and early Spring are the outbreak period of flu. The example of Austria peak weeks are around the end of the year (before and after Christmas time, 50 * 7 = 350). 

## Question 4 solution
For each country, find which year had the lowest mortality.

In [25]:
# Get columns from the original dataset == > data 
df2 = data[['location','deaths_2010_all_ages','deaths_2011_all_ages','deaths_2012_all_ages','deaths_2013_all_ages',
           'deaths_2014_all_ages','deaths_2015_all_ages','deaths_2016_all_ages','deaths_2017_all_ages',
           'deaths_2018_all_ages','deaths_2019_all_ages','deaths_2020_all_ages','deaths_2021_all_ages']]

In [26]:
# Fill missing values with 0 
df2 = df2.fillna(0)

In [27]:
# Sum up the mortality of each year of each country
df2 = df2.groupby('location').sum()

In [28]:
#reset the index (by default using ordered number index)
df2.reset_index(inplace=True)

In [29]:
# Take a look at the first 5 rows of df2
df2.head(5)

location  deaths_2010_all_ages  deaths_2011_all_ages  deaths_2012_all_ages  \
0   Austria               76998.0               76256.0               78996.0   
1   Belgium              104792.0              103919.0              108375.0   
2  Bulgaria              109235.0              108159.0              108362.0   
3    Canada              239635.0              242895.0              245240.0   
4     Chile                   0.0                   0.0                   0.0   

   deaths_2013_all_ages  deaths_2014_all_ages  deaths_2015_all_ages  \
0               79276.0               77986.0               81432.0   
1              108993.0              104324.0              110286.0   
2              103972.0              108592.0              110137.0   
3              251760.0              257735.0              263220.0   
4                   0.0                   0.0                   0.0   

   deaths_2016_all_ages  deaths_2017_all_ages  deaths_2018_all_ages  \
0               78842.0               81311.0               82063.0   
1              107532.0              109278.0              110341.0   
2              107015.0              109184.0              108202.0   
3              266080.0              277060.0              282915.0   
4              103375.0              106045.0              106458.0   

   deaths_2019_all_ages  deaths_2020_all_ages  deaths_2021_all_ages  
0               81485.0               91508.0                3885.0  
1              108436.0              128770.0                2423.0  
2              107818.0              126075.0                4932.0  
3              283065.0              240315.0                   0.0  
4              109079.0              127335.0                4666.0

In [30]:
# Extract the location column and put the data in a list
location = []
for i in df2['location']:
    location.append(i)

In [31]:
# Drop the column location (prepare for getting the lowest value of each row later)
df_data = df2.drop(columns=['location'])

In [32]:
# calculate the minimum value along each row of the DataFrame ==> df_data
# Function suggested by the link
# https://www.geeksforgeeks.org/how-to-get-the-minimum-value-from-the-pandas-dataframe-in-python/
min_rows = df_data.min(axis=1)

In [33]:
# Get the corresponding column name of the minimum value of each row
# Function suggested by the link
# https://www.geeksforgeeks.org/python-pandas-dataframe-idxmin/?ref=header_search
min_columns = df_data.idxmin(axis=1)

In [35]:
type(min_rows) # series data type

pandas.core.series.Series

In [36]:
type(min_columns) # series data type

pandas.core.series.Series

In [37]:
# Write a DataFrame includes 3 columns (location, Year, lowest_mortality)
result_df = pd.DataFrame({'location':location,'year': min_columns,'lowest_mortality': min_rows})

In [38]:
# Take a look at the result
result_df

location                  year  lowest_mortality
0            Austria  deaths_2021_all_ages            3885.0
1            Belgium  deaths_2021_all_ages            2423.0
2           Bulgaria  deaths_2021_all_ages            4932.0
3             Canada  deaths_2021_all_ages               0.0
4              Chile  deaths_2010_all_ages               0.0
5            Croatia  deaths_2021_all_ages               0.0
6            Czechia  deaths_2021_all_ages               0.0
7            Denmark  deaths_2021_all_ages            3703.0
8    England & Wales  deaths_2021_all_ages           35793.0
9            Estonia  deaths_2021_all_ages             751.0
10           Finland  deaths_2021_all_ages               0.0
11            France  deaths_2021_all_ages               0.0
12           Germany  deaths_2010_all_ages               0.0
13            Greece  deaths_2010_all_ages               0.0
14           Hungary  deaths_2021_all_ages               0.0
15           Iceland  deaths_2021_all_ages               0.0
16            Israel  deaths_2021_all_ages            1083.0
17             Italy  deaths_2010_all_ages               0.0
18            Latvia  deaths_2021_all_ages            1854.0
19         Lithuania  deaths_2021_all_ages            1179.0
20        Luxembourg  deaths_2021_all_ages               0.0
21       Netherlands  deaths_2021_all_ages           11733.0
22       New Zealand  deaths_2010_all_ages             538.0
23  Northern Ireland  deaths_2010_all_ages               0.0
24            Norway  deaths_2021_all_ages               0.0
25            Poland  deaths_2021_all_ages           10090.0
26          Portugal  deaths_2021_all_ages            3634.0
27          Scotland  deaths_2021_all_ages            4813.0
28          Slovakia  deaths_2021_all_ages               0.0
29          Slovenia  deaths_2021_all_ages               0.0
30       South Korea  deaths_2021_all_ages               0.0
31             Spain  deaths_2021_all_ages           21110.0
32            Sweden  deaths_2021_all_ages            2191.0
33       Switzerland  deaths_2021_all_ages               0.0
34            Taiwan  deaths_2021_all_ages               0.0
35    United Kingdom  deaths_2010_all_ages               0.0
36     United States  deaths_2010_all_ages               0.0

Cross check with the result from df2 and we can see for example the lowest value for Austria was from 2021 with a number of 3885, the same number of 3885 from result_df. The result_df also shows some places had the lowest 0 (this might be because of missing value). The lowest values occurred mostly in 2021 for each location and this may be due to the end of the COVID-19 period.